#18826543: API каталога

djpont opened the pull request:
Was done:
   - [x] Каталог с фильтром по цене и категориями.
   - [x] Страничка товара.
   - [x] Профиль и страничка редактирования профиля.
   - [x] Запрос профиля при открытии сайта.
   - [x] Линтер проверил, проект билдится.

djpont committed:
feat: store и api для Machineries

djpont committed:
feat: добавил в общий store

mrHoft commented on src/shared/store/machineries/hooks.ts:
```diff
@@ -0,0 +1,6 @@
+
+import { getMachineries } from './slice';
+
+export const useGetMachineries = (): TUserDto | null => getMachineries(store.getState());
```
Тут тип должен быть `TMachineryDto[]`

mrHoft commented on src/shared/store/machineries/thunks.ts:
```diff
@@ -0,0 +1,20 @@
+import { machineriesAPI, type ValidationErrors } from '@/shared/api';
+
+export const fetchGetMachineries = createAsyncThunk(
+  'session/getUser',
```
Не верное наименование для Machineries

mrHoft reviewed (approved):
В целом, вижу ты идею понял. Отлично!

djpont committed:
feat: ручки каталога

Каталог с фильтром по цене и категориями.
Страничка товара.
Профиль и страничка редактирования профила.
Запрос профиля при открытии сайта.
Линтер проверил, проект билдится.

djpont committed:
feat: заглушка

djpont committed:
feat: заголовок

mrHoft commented on src/app/catalog/[id]/page.tsx:
```diff
@@ -1,20 +1,15 @@
-  const { id } = params;
-  console.log(id);
-  //const itemData = fetchProductDataById(id);
+  const itemData = await fetch(`${BASE_URL}/${API_ENDPOINTS.machinery}/${id}`).then(data =>
```
Я помню что ты писал о затруднениях связанных со сборкой и переносом запросов на конкретные страницы...
Но это... Так нельзя делать. Ты прям жёстко "положил" на архитектуру приложения :)
Сделай хук в `api/machinery` и вызывай его чтоб получить itemData. Тем более что у тебя там уже есть getMachineries!

djpont replied:
Не получилось, у нас axios для запросов использует localStogare, на сервере его нет.
getMachineries это стор, его тоже на сервере нет. Это ssr страничка.

djpont replied:
У меня есть getMachinery для этой странички специально, но оно только на клиенте работает в таком случае.

mrHoft replied:
getMachinery делает то же самое только через axios. Странно. Но вот объяснение:
> In addition to fetch() on the client-side, Next.js polyfills fetch() in the Node.js environment. You can use fetch() in your server code (such as getStaticProps/getServerSideProps)

Но можно "вручную" сделать так чтоб axios работал с Next. Тут пример:
https://reacthustle.com/blog/how-to-use-axios-with-nextjs-using-axios-hooks

djpont replied:
Сделал через getMachinery, для ручек техники использую axios.get

djpont replied:
В общем странички техники (конкретной) у нас рендерятся на севере и полный ssr!
А вот каталог нет, увы.

djpont replied:
Есть решение - сделать категории не через url-параметр, а через роутинг некста, то есть
http://localhost:3000/catalog/1 - трактора
http://localhost:3000/catalog/5 - сеялки
А странички товаров глубже сделать
http://localhost:3000/catalog/1/3 - трактор с id=3

djpont replied:
Получилось!)

mrHoft commented on src/app/layout.tsx:
```diff
@@ -11,38 +11,33 @@ import { ModalProvider } from '@/entities/Modal';
-          {/* </StoreProvider> */}
+          <StoreProvider store={store}>
+            <PageContent>
+              <UserProvider>
```
Нужно будет не забыть потом убрать это, учитывая что ты если не везде то уже много где убрал useUserContext.

mrHoft commented on src/shared/catalog/api/get.ts:
```diff
@@ -1,21 +1,26 @@
       switch (key) {
-        case 'category':
-          return !value || item.category === value;
+        // case 'category':
+        //   return !value || item.machinery.category === value;
```
Считаю не достойным докапываться до комментов, но тут если и оставил "верезанный" кейс то стоило хотя бы написать коммент почему он не используется. Я то помню из обсуждения а для стороннего "читателя" это выглядит как "не прибранный мусор".

djpont replied:
Понимаю, но у меня турбо-режим с пониженным качеством кода)))

mrHoft reviewed (approved):
Серьезно ты подошёл к задаче: ни на кого уже больше не рассчитывая решил закрыть все явные дыры.

Обалдеть ты заморочался с MetaData!
Вроде делал machinerts, а даже user-а сделал!

Вот это понравилось:
```ts
export const useAuth = (): boolean => {
  const user = useGetUser();

  return !(!user && typeof window !== 'undefined');
};
```

djpont committed:
feat: вместо instance использую напрямую axios.get, теперь страницы техники и каталога ssr.

djpont committed:
feat: поправил url

djpont committed:
feat: хук теперь смотрит /название-категории/

djpont committed:
feat: хук теперь смотрит /название-категории/

djpont committed:
feat: url теперь выглядит так /название-категории/, добавил ssr

djpont committed:
feat: url теперь выглядит так /название-категории/

djpont committed:
feat: url теперь выглядит так /название-категории/

djpont committed:
feat: добавил ручку для главной страницы каталога

djpont committed:
feat: поправил url

djpont committed:
feat: категория теперь словом, а не число

mrHoft reviewed (approved):
Жаль сразу не догадались что не ко всем ручкам нужен одинаковый подход.
Отличная реализация!

djpont committed:
#19844953: API Заказов (#116)

Was done:
   - [x] Orders API и Store
   - [x] Оформление заказа
   - [x] Текущие заказы в профиле
   - [x] История заказов в профиле
   - [x] Линтер проверил, проект билдится.
